### ISSUE TITLE
Fix S2583 FP: ++ operator not taken into account when evaluating modulo operations

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#8517]
### Description

It seems that the ++ operator is not taken into account when evaluating modulo operations.

### Repro steps

The following claims always to evaluate to `true`.
``` C#
public int PlusPlusTakenIntoAccount(string str)
{
    for (var index = 0; index < str.Length; index++)
    {
        if (index % 4 == 0) // Compliant.
        {
            return 42;
        }
    }
    return 0;
}
```

When using a while statement, the expression is claimed always to evaluate to `false`.

``` C#
 public int plusPlusNotTakenIntoAccount(string str)
 {
     var index = 0;
     while (index++ < str.Length)
     {
         if (index % 4 == 0) // Compliant.
         {
             return 42;
         }
     }
     return 0;
 }
```

Changing the index in a seperate statement does not change anything:

``` C#
 public int plusPlusNotTakenIntoAccount(string str)
 {
     var index = 0;
     while (index < str.Length)
     {
         index++;
         if (index % 4 == 0) // Compliant.
         {
             return 42;
         }
     }
     return 0;
 }
 ```

If this code would be applied on a string of a known value, we can be certain of the outcome.

``` C#
public int plusPlusNotTakenIntoAccountConst()
{
    var str = "42";
    var index = 0;
    while (index < str.Length)
    {
        index++;
        if (index % 4 == 0) // Noncompliant, index will be between 1 and 3.
        {
            return 42;
        }
    }
    return 0;
}
```

I detected this, while writing code to format an IBAN, (hence the mod 4). If more info is needed, please let me know.

Obliviously, the latest version of Sonar was used.

**Comment by gregory-paidis-sonarsource** [comment]
Hey Corniel, thanks for the report.
This is part of a broader problem around loop handling in our symbolic execution engine.
It will be fixed as part of #8474. 
Feel free to track the progress there.

**Comment by Corniel** [comment]
@gregory-paidis-sonarsource Any ETA for this?
